Apply ZIO SBT Ecosystem and CI Plugins

khajavi opened the pull request.

khajavi committed:
initial work.

khajavi committed:
core module.

khajavi committed:
core tests module.

khajavi committed:
laws module.

khajavi committed:
macros module.

khajavi committed:
experimental module.

khajavi committed:
experimental-laws module.

khajavi committed:
experimental tests module.

khajavi committed:
scala-parallel-collections module.

khajavi committed:
other modules.

khajavi committed:
fix cross scala versions.

khajavi committed:
make +publishLocal works properly.

khajavi committed:
remove build helper.

khajavi committed:
remove website workflow.

khajavi committed:
remove BuildHelper completely.

khajavi committed:
generate ci workflow.

khajavi committed:
add java options.

khajavi committed:
add scalajs scalac option.

khajavi committed:
move scalajs to jsSettings.

khajavi committed:
remove experimental native tests

khajavi committed:
remove scalajs option from experimentalTestJS module.

khajavi committed:
sequential tests.

khajavi committed:
add scalajs options.

khajavi committed:
revert to scalajs 1.12.0

khajavi committed:
use scala3 dialect for scalafmt.

khajavi committed:
remove extra commands.

khajavi committed:
remove extra zio version.

khajavi committed:
Revert "remove extra zio version."

This reverts commit e64044e446f40dd75c78079b50dd4c73dd39ecea.

khajavi committed:
add support for all major jvm platforms.

khajavi committed:
remove extra plugins.

khajavi committed:
Revert "use scala3 dialect for scalafmt."

This reverts commit 0b607c61

sideeffffect commented:
I'm all for improving our CI/sbt story (which I know currently is less than ideal)... But wow, the number of GitHub Actions jobs is _huge_! I'm worried that it wouldn't be reasonable to have such high number, each PR would be _very_ expensive and take a long time to complete.
Perhaps there could be a way to reduce the number. Maybe this can serve as some sort of inspiration https://en.wikipedia.org/wiki/Latin_hypercube_sampling

Or keep the original idea where JVM 17, Scala 2.13 and JVM platform are the default and we diverge from it only in one axis.

khajavi commented:
@sideeffffect 
With zio-sbt, we can change the granularity of tests. Currently, I need more granularity to find errors. Then I will make tests sequential.

khajavi committed:
add zio test.

khajavi committed:
fix linking errors.

khajavi committed:
fix linking errors.

khajavi committed:
use sequential tests.

khajavi committed:
increase heap size.

khajavi committed:
update ci.

khajavi committed:
display free size.

khajavi committed:
scalafix.

khajavi committed:
remove extra char.

khajavi committed:
exclude scala3 sources from fmt.

sideeffffect reviewed (commented):
We still want to keep 2.11.

khajavi commented:
> We still want to keep 2.11.

Ok, I need to override the `sbt-scalajs` plugin. Did you find out how to do that?

khajavi committed:
increase swap size.

khajavi committed:
override runner for scala3 sources.

khajavi committed:
show memory and swap size.

adamgfraser commented:
Can we not show the memory? I think it is extremely disruptive in tests and a symptom of other problems that need to be handled separately.

khajavi commented:
> Can we not show the memory? I think it is extremely disruptive in tests and a symptom of other problems that need to be handled separately.

Yes, I added it for debugging purposes to find out the proper size for heap and swap size. Now that the tests are green, I can remove them.

adamgfraser commented:
Okay. The other thing we should be doing here is making sure we are running each test file and each project sequentially. We already run tests in parallel so there is no point in also running different test files or projects in parallel and it can create additional memory pressure that leads to test failures or decreased test performance. In this project I added a couple of command aliases to do that.

khajavi committed:
revert to only java 17.

khajavi committed:
remove memory footprint.

khajavi committed:
install libuv.

khajavi committed:
java11 support for ci tests.

khajavi committed:
resurrect scala 2.11

khajavi committed:
remove extra project aggregations.

I assumed they were only used in ci.

khajavi committed:
clean up.

khajavi committed:
remove extra dependency.

khajavi committed:
enable native test on experimentalTests module.

khajavi committed:
drop scala 2.11 from benchmarks module.

khajavi committed:
scalafmt.

khajavi committed:
stick with sbt-scala-native 0.4.10 to support scala 2.11

khajavi committed:
limit parallel tests to 3.

We need to monitor test behaviour to find out the proper size.

khajavi committed:
update zio-sbt plugin.

This includes +publishLocal and +Test/compile

khajavi committed:
publish for java 8.

khajavi committed:
suppress scaladoc's linking problem.

khajavi committed:
refactor.

khajavi commented:
> Okay. The other thing we should be doing here is making sure we are running each test file and each project sequentially. We already run tests in parallel so there is no point in also running different test files or projects in parallel and it can create additional memory pressure that leads to test failures or decreased test performance. In this project I added a couple of command aliases to do that.

I added an option to limit the parallel task on the matrix strategy. Also, grouped tests together based on scala and java versions, to make them more sequential.

adamgfraser commented:
@khajavi So it seems like this has resulted in a dramatic increase in CI times. It is taking around 20 minutes for these tests now versus it previously took around 4 minutes after we merged my PR upgrading the ZIO version. Can we please maintain the current execution semantics?

khajavi commented:
@adamgfraser Okay, I will try to reduce the duration of the CI process. If this does not lead to any improvement, we need to consider discontinuing automatic test generation.

khajavi committed:
only test on java 17 platform.

khajavi committed:
set max metaspace size to 4G.

khajavi commented:
I'll update this after each try:
- After setting to only test on Java 17=> decreased from 1h 28m to 1h 1m
- After setting max metaspace size to 4G => decreased to 1h and 4m

khajavi committed:
disable checking workflow.

khajavi committed:
remove jvm options from sbt command.

khajavi committed:
cache dependencies.

khajavi committed:
generate tests using aggregate projects.

khajavi committed:
customize ci.

khajavi committed:
fix mistake on scala version.

khajavi committed:
rewrite the current ci to zio-sbt-ci plugin.

khajavi committed:
remove matrix strategy for publishLocal.

khajavi committed:
reorder dependencies.

khajavi committed:
test benchmark module compilation.

khajavi commented:
@adamgfraser To accomplish this task, I utilized the existing semantic, as you said, and attempted to migrate it to the zio-sbt plugin. This involved applying mostly the same workflows used in the original code to the version.

By doing so, I expect to have the same level of quality as the current workflow execution.

khajavi committed:
Merge branch 'series/2.x' into apply-zio-ecosystem-plugin

khajavi committed:
update zio-sbt plugins.

guizmaii commented:
@khajavi, Could you update this PR, maybe, please? I'd be interested in merging these changes 🙂

khajavi commented:
@guizmaii 
I'm currently engaged in some other things, but I might revisit this later.
Also, I think this PR is open to any collaboration.
